rules/maximum-lines-in-function.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * @fileoverview Set a maximum number of lines per function
+ * @author Beau Gunderson <beau@beaugunderson.com>
+ */
+
+"use strict";
+
+module.exports = {
+	meta: {
+		docs: {
+			recommended: false,
+			type: "error",
+			description: "Set a maximum number of lines per function"
+		},
+
+		schema: [
+			{
+				type: "number"
+			}
+		]
+	},
+
+	create: function(context) {
+		var maximumLines = context.options ? context.options : 25;
+		var sourceCode = context.getSourceCode();
+
+		function inspectFunctionDeclaration(emitted) {
+			if (emitted.exit) {
+				return;
+			}
+
+			var node = emitted.node;
+
+			var startingLine = sourceCode.getLine(node);
+			var endingLine = sourceCode.getEndingLine(node);
```

Comment:
I'm slightly confused about this rule. Does it mean to enforce the number of lines inside the function?
Regardless of whether they are blank lines or stuffed with solidity statements (maybe even multiple statements per line)?
OR does it mean to check for the number of solidity statements inside this function?

So if my configuration is to allow `2` lines, SHOULD the below qualify:

```
function foo() {
    // plenty of blank lines below
    CallFunction("hello);




}
```

```
function bar() {
    a(); b(); c(); d(); e(); f();
}
```

Reply:
I think the goal is to limit the number of statements, lines is usually (but not always) a good proxy for that; I'll see if I can figure out a better way that doesn't include blank lines in the count. :)

Reply:
it'll be pretty simply. Just examine the `FunctionDeclaration` node you're catching inside the rule. its `body` contains a `BlockStatement` node whose `body` attribute is a List of all statement nodes inside the function block. You will probably need to use the length of this list and a count of newlines I guess?

Reply:
ah, easier than I thought! I'll implement that now

Reply:
interestingly, for this function I get 10 top-level statements:

```
    function insertOrderIntoList(Order.Data storage _order, bytes32 _betterOrderId, bytes32 _worseOrderId) private returns (bool) {
        bytes32 _bestOrderId = bestOrder[getBestOrderWorstOrderHash(_order.market, _order.outcome, _order.orderType)];
        bytes32 _worstOrderId = worstOrder[getBestOrderWorstOrderHash(_order.market, _order.outcome, _order.orderType)];
        IOrdersFetcher _ordersFetcher = IOrdersFetcher(controller.lookup("OrdersFetcher"));
        (_betterOrderId, _worseOrderId) = _ordersFetcher.findBoundingOrders(_order.orderType, _order.price, _bestOrderId, _worstOrderId, _betterOrderId, _worseOrderId);
        if (_order.orderType == Order.Types.Bid) {
            _bestOrderId = updateBestBidOrder(_order.id, _order.market, _order.price, _order.outcome);
            _worstOrderId = updateWorstBidOrder(_order.id, _order.market, _order.price, _order.outcome);
        } else {
            _bestOrderId = updateBestAskOrder(_order.id, _order.market, _order.price, _order.outcome);
            _worstOrderId = updateWorstAskOrder(_order.id, _order.market, _order.price, _order.outcome);
        }
        if (_bestOrderId == _order.id) {
            _betterOrderId = bytes32(0);
        }
        if (_worstOrderId == _order.id) {
            _worseOrderId = bytes32(0);
        }
        if (_betterOrderId != bytes32(0)) {
            orders[_betterOrderId].worseOrderId = _order.id;
            _order.betterOrderId = _betterOrderId;
        }
        if (_worseOrderId != bytes32(0)) {
            orders[_worseOrderId].betterOrderId = _order.id;
            _order.worseOrderId = _worseOrderId;
        }
        return true;
    }
```

Reply:
which I think are the 4 declarations, the 5 `if`s, and the `return`

Reply:
Ah! Okay 10 is expected here (as you said, top-level). Okay:
- Take all 10 (top level) items in the list
- For each item, fetch its starting line & ending line

This info can be used to easily calculate total number of lines which contain some code, which is acceptable instead of the exact number of statements.

Reply:
@duaraghav8 perfect, implemented that and renamed the rule to `maximum-statements-in-function`; I now get 26 for the function above (which I believe is the correct behavior)